Initium-Odp/src/com/universeprojects/miniup/server/aspects/AspectSlotted.java

```diff
@@ -1,40 +1,197 @@
 	@Override
 	public List<ItemPopupEntry> getItemPopupEntries(CachedEntity currentCharacter)
 	{
-		return null;
+		if (CommonChecks.checkItemIsAccessible(object.getEntity(), currentCharacter) && hasOpenSlots()) {
+			ItemPopupEntry ipe = new ItemPopupEntry("Insert an item into the slots", "Insert a slottable item into this item.",
+					"viewSlottableItems(event, " + object.getKey().getId() + ")");
+			return Arrays.asList(new ItemPopupEntry[]
+			{
+					ipe
+			});
+		}
+		else return null;
 	}
 
 	@Override
-	public String getPopupTag()
-	{
+	public String getPopupTag() {
 		if (getMaxCount()>0)
 			return "Slotted";
 		else
 			return null;
 	}
 
-	
-	public long getMaxCount()
-	{
-		Long count = (Long)this.getProperty("maxCount");
+	/**
+	 * Returns the number of max slots on this aspect.
+	 * @return
+	 */
+	public long getMaxCount() {
+		Long count = (Long)getProperty("maxCount");
 		if (count==null) count = 0L;
 		return count;
 	}
 	
-}
+	/**
+	 * Returns true if this aspect has any open slots.
+	 * @return
+	 */
+	public boolean hasOpenSlots() {
+		return getMaxCount() < getOpenCount();
+	}
+	
+	/**
+	 * Returns the number of open slots on this aspect.
+	 * @return
+	 */
+	public long getOpenCount() {
+		Long max = (Long)getProperty("maxCount");
+		
+		List<EmbeddedEntity> currentlySlotted = (List<EmbeddedEntity>)getProperty("slotItems"));
+		
+		if(currentlySlotted == null) return max;
+		
+		return max-currentlySlotted.size();
+	}
+	
+	static {
+		addCommand("InsertItemToSlot", AspectSlotted.CommandInsertItemToSlot.class);
+	}
+	
+	
+	/**
+	 * given a slotted item and a slottable item, insert it into the slot and apply the
+	 * appropriate modifiers to the parent item.
+	 * @author Evan
+	 *
+	 */
+	public static class CommandInsertItemToSlot extends Command {
+
+		public CommandInsertItemToSlot(ODPDBAccess db, HttpServletRequest request, HttpServletResponse response) {
+			super(db, request, response);
+		}
+		
+		//Pass the ID of the item with slots and the item we're slotting in. We will need to sanitize it.
+		@Override
+		public void run(Map<String, String> parameters) throws UserErrorMessage, UserRequestIncompleteException{
+			
+			//generate the base item from the given ID, then do basic checks.
+			Key baseKey = KeyFactory.createKey("Item", Long.parseLong(parameters.get("baseId")));
+			InitiumObject baseItem = db.getInitiumObject(baseKey);
+			
+			if(baseItem == null) throw new UserErrorMessage("Invalid Item ID.");
+			if(!CommonChecks.checkItemIsAccessible(baseItem.getEntity(), db.getCurrentCharacter()))
+				throw new UserErrorMessage("Nice try.");
+			if(!CommonChecks.checkItemHasAspect(baseItem.getEntity(), "Slotted"))
+				throw new UserErrorMessage("This item doesn't have any slots.");
+			
+			//gets a compiler error in ODP environment but should work when deployed.
+			AspectSlotted slottedAspect = (AspectSlotted) baseItem.getInitiumAspect("Slotted");
```

Comment:
There shouldn't be compile errors in ODP, what is the error?

Reply:
I figured it out, the supertype of GameObject is missing from the ODP so I get some weird errors. I copied this code from the aspect pet so it will be fine.

Reply:
GameObject is an interface that comes from the GEF-Common project. Do you not get that one when you setup?

Reply:
Nope, as far as I can tell the GEF-Common project isn't on Github.